### ISSUE TITLE
Update @deprecated to work with properties and __init__

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8124]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8124|
|Type|enhancement|
|Created|2015-11-23 09:31:41Z|
|Branch|https://github.com/twisted/twisted/tree/deprecated-property-8124|

from glyph

If you care about inheritance, the implementation is a little bit tricky, because you have to manually walk the class hierarchy looking for the attribute.  But conceptually it's pretty simple: just look at the type of 'oself' in the __get__ method of the returned descriptor.

Right now, @deprecated is hard-coded to assume a function, but it could be repurposed to work with specific other descriptor types reasonably easily.  Arbitrary descriptors might be hard, because it's not clear when to emit the message, but specific types like @property should be pretty straightforward with an instance check.

<details><summary>Searchable metadata</summary>

```
trac-id__8124 8124
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__branches_deprecated_property_8124 branches/deprecated-property-8124
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1448271101384950 1448271101384950
changetime__1458100479509165 1458100479509165
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

(In [46325]) Branching to deprecated-property-8124.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @hawkowl|
|-|-|

I am lost and Amber said she might be able to help here.

Test are in the branch :)

You can check if the tests are valid and if they show correct usage of @deprecated

If you know how to update the code to have the tests pass please commit to this branch.

Also feel free to update the tests is they don't make sense.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Thanks Adi, the tests were a good starting point -- I've updated it with how ```@deprecated``` is used on classes existingly -- as a class decorator, not decorating init.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl removed owner|
|-|-|

I've added some implementations -- I couldn't get it to work with just @deprecated, as it didn't handle the case where you deprecated the *property* -- ```@wraps``` just completely falls apart. So, I deprecate the inner method which is a function -- not a method, so the existing FQDN lookup stuff doesn't know it's on a class!, and because we know it's a property, we can stack walk and probably get the right answer.

So yes, please review. Builders are spun.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for your update.

I have updated the test to have all decorated related tests into a single testcase.

I have also updated them to check for the deprecatedVersion attribute which is set by the decorator... but I have no idea why we need it.

I have also updated the test to include a test for the setter decorator.

I then tried an ugly change which will only require a single decorator so we can have something like this

```
    @deprecatedProperty(Version('Twisted', 1, 2, 3))
    def someProperty(self):
        """
        Getter docstring.
        """
        return self._someProtectedValue


    @someProperty.setter
    def someProperty(self, value):
        """
        Setter docstring.
        """
        self._someProtectedValue = value
```

Amber, please check my latest commits.

Feel free to revert or improve them. I am not happy with deprecatedFunction code duplication but I got tired of fighting Python build-in magic. 


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

They look fine, but since we've both worked on this, we'll need someone else to review it :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks adi, hawkie.

[Appease twistedchecker, lest its terrible wrath destroy us all](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/3696/steps/run-twistedchecker/logs/new%20twistedchecker%20errors), then land.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Something I noticed:

https://github.com/twisted/twisted/compare/trunk...deprecated-property-8124#diff-29059ef15a88640b2a2751c2a58a938dR800 says "test_classMethod" but we're not actually testing classmethods, it's actually a class. So rename the test method to "test_class" instead I guess?

After that, and the twistedchecker error fixed up, please land, Adi :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set status to `closed`|
|-|-|

(In [47001]) Merge deprecated-property-8124: Make @deprecatedProperty, to make deprecated properties

Author: adiroiban, hawkowl
Reviewer: glyph
Fixes: [#8124](https://github.com/twisted/twisted/issues/8124)
